File: examples/javascript/twit_api_client.js
#!/usr/bin/env node
// TWiT API Client
// A more comprehensive client for interacting with the TWiT.tv API

const https = require('https');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

// Simple YAML parser for our basic needs
function parseYAML(content) {
  const lines = content.split('\n');
  const result = {};
  let currentSection = null;
  
  for (const line of lines) {
    // Skip comments and empty lines
    if (line.trim().startsWith('#') || !line.trim()) continue;
    
    // Handle section headers (keys with no value)
    if (!line.startsWith(' ') && line.includes(':')) {
      const parts = line.split(':');
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      
      if (!value) {
        currentSection = key;
        result[currentSection] = {};
      } else {
        // Remove any inline comments from the value
        const valueParts = value.split('#');
        result[key] = valueParts[0].trim();
      }
    } 
    // Handle indented properties (section values)
    else if (currentSection && line.trim().includes(':')) {
      const parts = line.trim().split(':');
      const key = parts[0].trim();
      const value = parts.slice(1).join(':').trim();
      
      // Remove any inline comments from the value
      const valueParts = value.split('#');
      result[currentSection][key] = valueParts[0].trim();
    }
  }
  
  return result;
}

class TwitApiClient {
  constructor(appId = null, appKey = null, baseUrl = null) {
    // Setup logger
    this.logger = {
      debug: (message) => console.debug(`[DEBUG] ${message}`),
      info: (message) => console.info(`[INFO] ${message}`),
      error: (message) => console.error(`[ERROR] ${message}`)
    };

    // Try to load credentials from file if not provided
    if (!appId || !appKey) {
      this.loadCredentialsFromFile();
    } else {
      this.appId = appId;
      this.appKey = appKey;
      this.baseUrl = baseUrl || 'https://twit.tv/api/v1.0';
    }
  }

  // Get a list of all shows
  async getShows(params = {}) {
    return this.makeRequest('/shows', params);
  }

  // Get a specific show by ID
  async getShow(id, params = {}) {
    return this.makeRequest(`/shows/${id}`, params);
  }

  // Get a list of all episodes
  async getEpisodes(params = {}) {
    return this.makeRequest('/episodes', params);
  }

  // Get a specific episode by ID
  async getEpisode(id, params = {}) {
    return this.makeRequest(`/episodes/${id}`, params);
  }

  // Get live streams information
  async getStreams(params = {}) {
    return this.makeRequest('/streams', params);
  }

  // Get people information
  async getPeople(params = {}) {
    return this.makeRequest('/people', params);
  }

  // Load credentials from the shared credentials file
  loadCredentialsFromFile() {
    const scriptDir = path.dirname(require.main ? require.main.filename : __filename);
    const credentialsFile = path.join(scriptDir, '..', 'credentials.yml');
    
    if (fs.existsSync(credentialsFile)) {
      try {
        const content = fs.readFileSync(credentialsFile, 'utf8');
        const credentials = parseYAML(content);
        this.appId = credentials.twit_api.app_id;
        this.appKey = credentials.twit_api.app_key;
        this.baseUrl = credentials.twit_api.base_url || 'https://twit.tv/api/v1.0';
        
        this.logger.debug(`Loaded credentials - APP_ID: ${this.appId}, BASE_URL: ${this.baseUrl}`);
      } catch (e) {
        this.logger.error(`Error loading credentials: ${e.message}`);
        throw new Error(`Failed to load credentials: ${e.message}. Please make sure your credentials.yml file is properly formatted.`);
      }
    } else {
      throw new Error(`Credentials file not found: ${credentialsFile}. Please copy credentials.yml.sample to credentials.yml and update with your actual credentials.`);
    }
  }

  // Make an HTTP request to the TWiT API
  makeRequest(endpoint, params = {}) {
    return new Promise((resolve, reject) => {
      try {
        // Build URL with query parameters if any
        let url = `${this.baseUrl}${endpoint}`;
        
        // Add query parameters if any
        if (Object.keys(params).length > 0) {
          const queryParams = new URLSearchParams();
          for (const key in params) {
            queryParams.append(key, params[key]);
          }
          url += `?${queryParams.toString()}`;
        }
        
        const parsedUrl = new URL(url);
        
        // Create options for request
        const options = {
          hostname: parsedUrl.hostname,
          path: parsedUrl.pathname + parsedUrl.search,
          method: 'GET',
          headers: {
            'Accept': 'application/json',
            'app-id': this.appId,
            'app-key': this.appKey
          }
        };
        
        this.logger.debug(`Making request to: ${url}`);
        
        const req = https.request(options, (res) => {
          let data = '';
          
          res.on('data', (chunk) => {
            data += chunk;
          });
          
          res.on('end', () => {
            const status = res.statusCode;
            
            if (status === 200) {
              try {
                const jsonData = JSON.parse(data);
                resolve(jsonData);
              } catch (e) {
                this.logger.error(`Error parsing JSON response: ${e.message}`);
                reject(new Error(`Error parsing JSON response: ${e.message}`));
              }
            } else if (status === 401 || status === 403) {
              this.logger.error("Authentication error: Check your app-id and app-key");
              resolve({ "error": "Authentication failed", "code": status });
            } else if (status === 404) {
              this.logger.error(`Resource not found: ${endpoint}`);
              resolve({ "error": "Resource not found. Check that the API endpoint is correct.", "code": status });
            } else if (status === 500) {
              if (data.includes('usage limits are exceeded')) {
                this.logger.error("API usage limits exceeded");
                resolve({ "error": "API usage limits exceeded", "code": status });
              } else {
                this.logger.error(`Server error: ${data}`);
                resolve({ "error": "Server error", "code": status });
              }
            } else {
              this.logger.error(`Unexpected response: ${status} - ${data}`);
              resolve({ "error": "Unexpected response", "code": status, "body": data });
            }
          });
        });
        
        req.on('error', (e) => {
          this.logger.error(`Request failed: ${e.message}`);
          resolve({ "error": e.message });
        });
        
        req.end();
      } catch (e) {
        this.logger.error(`Error creating request: ${e.message}`);
        resolve({ "error": `Error creating request: ${e.message}` });
      }
    });
  }
}

// Example usage
if (require.main === module) {
  (async () => {
    try {
      const client = new TwitApiClient();
      
      console.log("Testing connection to TWiT API...");
      
      // Test getting shows
      const result = await client.getShows();
      
      if (result.error) {
        console.log(`Error: ${result.error}`);
      } else {
        console.log("Connection successful!");
        console.log(`Found ${result.count} shows`);
        
        if (result.shows && result.shows.length > 0) {
          console.log("\nFirst few shows:");
          result.shows.slice(0, 3).forEach(show => {
            console.log(`- ${show.label} (${show.id})`);
          });
          
          // Try to get a specific show (using the first show's ID if available)
          const firstShowId = result.shows[0].id;
          console.log(`\nGetting details for show ID: ${firstShowId}`);
          
          const showDetails = await client.getShow(firstShowId);
          if (!showDetails.error) {
            // For single show, the response has a 'shows' object, not an array
            if (showDetails.shows) {
              const show = showDetails.shows;
              console.log(`Show title: ${show.label}`);
              if (show.description) {
                console.log(`Description: ${show.description.substring(0, 100)}...`);
              }
            }
          }
        }
        
        // Get streams information
        console.log("\nGetting live streams information...");
        const streams = await client.getStreams();
        if (!streams.error && streams.streams) {
          console.log(`Found ${streams.count} streams`);
          streams.streams.slice(0, 2).forEach(stream => {
            console.log(`- ${stream.label} (${stream.streamType})`);
          });
        }
      }
    } catch (e) {
      console.error(`Error: ${e.message}`);
    }
  })();
}

module.exports = TwitApiClient;

File: examples/python/twit_api_client.py
#!/usr/bin/env python
# TWiT API Client
# A more comprehensive client for interacting with the TWiT.tv API

import requests
import json
import logging
import yaml
import os
import sys

class TwitApiClient:
    def __init__(self, app_id=None, app_key=None, base_url=None):
        # Try to load credentials from file if not provided
        if app_id is None or app_key is None:
            self.load_credentials_from_file()
        else:
            self.app_id = app_id
            self.app_key = app_key
            self.base_url = base_url or 'https://twit.tv/api/v1.0'
        
        # Setup logger
        self.logger = logging.getLogger('twit_api')
        handler = logging.StreamHandler()
        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        handler.setFormatter(formatter)
        self.logger.addHandler(handler)
        self.logger.setLevel(logging.INFO)

    # Get a list of all shows
    def get_shows(self, params=None):
        return self.make_request('/shows', params or {})

    # Get a specific show by ID
    def get_show(self, id, params=None):
        return self.make_request(f'/shows/{id}', params or {})

    # Get a list of all episodes
    def get_episodes(self, params=None):
        return self.make_request('/episodes', params or {})

    # Get a specific episode by ID
    def get_episode(self, id, params=None):
        return self.make_request(f'/episodes/{id}', params or {})

    # Get live streams information
    def get_streams(self, params=None):
        return self.make_request('/streams', params or {})

    # Get people information
    def get_people(self, params=None):
        return self.make_request('/people', params or {})

    # Load credentials from the shared credentials file
    def load_credentials_from_file(self):
        script_dir = os.path.dirname(os.path.abspath(__file__))
        credentials_file = os.path.join(script_dir, '..', 'credentials.yml')
        
        if os.path.exists(credentials_file):
            try:
                with open(credentials_file, 'r') as file:
                    credentials = yaml.safe_load(file)
                self.app_id = credentials['twit_api']['app_id']
                self.app_key = credentials['twit_api']['app_key']
                self.base_url = credentials['twit_api'].get('base_url', 'https://twit.tv/api/v1.0')
            except Exception as e:
                self.logger.error(f"Error loading credentials: {e}")
                raise Exception(f"Failed to load credentials: {e}. Please make sure your credentials.yml file is properly formatted.")
        else:
            raise FileNotFoundError(f"Credentials file not found: {credentials_file}. Please copy credentials.yml.sample to credentials.yml and update with your actual credentials.")

    # Make an HTTP request to the TWiT API
    def make_request(self, endpoint, params=None):
        # Build URL with query parameters if any
        url = f"{self.base_url}{endpoint}"
        
        # Create headers
        headers = {
            'Accept': 'application/json',
            'app-id': self.app_id,
            'app-key': self.app_key
        }
        
        self.logger.debug(f"Making request to: {url}")
        
        try:
            response = requests.get(url, headers=headers, params=params)
            
            if response.status_code == 200:
                return response.json()
            elif response.status_code in (401, 403):
                self.logger.error("Authentication error: Check your app-id and app-key")
                return {"error": "Authentication failed", "code": response.status_code}
            elif response.status_code == 404:
                self.logger.error(f"Resource not found: {endpoint}")
                return {"error": "Resource not found", "code": response.status_code}
            elif response.status_code == 500:
                if 'usage limits are exceeded' in response.text:
                    self.logger.error("API usage limits exceeded")
                    return {"error": "API usage limits exceeded", "code": response.status_code}
                else:
                    self.logger.error(f"Server error: {response.text}")
                    return {"error": "Server error", "code": response.status_code}
            else:
                self.logger.error(f"Unexpected response: {response.status_code} - {response.text}")
                return {"error": "Unexpected response", "code": response.status_code, "body": response.text}
        except Exception as e:
            self.logger.error(f"Request failed: {e}")
            return {"error": str(e)}


# Example usage
if __name__ == "__main__":
    try:
        client = TwitApiClient()
        
        print("Testing connection to TWiT API...")
        
        # Test getting shows
        result = client.get_shows()
        
        if "error" in result:
            print(f"Error: {result['error']}")
        else:
            print("Connection successful!")
            print(f"Found {result['count']} shows")
            
            if 'shows' in result and result['shows']:
                print("\nFirst few shows:")
                for show in result['shows'][:3]:
                    print(f"- {show['label']} ({show['id']})")
            
            # Try to get a specific show (using the first show's ID if available)
            if 'shows' in result and result['shows']:
                first_show_id = result['shows'][0]['id']
                print(f"\nGetting details for show ID: {first_show_id}")
                
                show_details = client.get_show(first_show_id)
                if "error" not in show_details:
                    # For single show, the response has a 'shows' object, not an array
                    if 'shows' in show_details:
                        show = show_details['shows']
                        print(f"Show title: {show['label']}")
                        if 'description' in show:
                            print(f"Description: {show['description'][:100]}...")
            
            # Get streams information
            print("\nGetting live streams information...")
            streams = client.get_streams()
            if "error" not in streams and 'streams' in streams:
                print(f"Found {streams['count']} streams")
                for stream in streams['streams'][:2]:
                    print(f"- {stream['label']} ({stream['streamType']})")
    except Exception as e:
        print(f"Error: {e}")
